File: backend/app/schemas/analytics.py
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field


class MonthlyCashflowItem(BaseModel):
    period: str
    income: int
    expense: int
    transfer: int
    net_cashflow: int
    savings_rate: float | None


class MonthlyCashflowResponse(BaseModel):
    items: list[MonthlyCashflowItem]


class CategoryMoMItem(BaseModel):
    period: str
    previous_period: str
    category: str
    current_amount: int
    previous_amount: int
    delta_amount: int
    delta_pct: float | None


class CategoryMoMResponse(BaseModel):
    items: list[CategoryMoMItem]
    reference_date: date | None = None
    is_partial_period: bool = False
    comparison_basis: Literal["same_day_previous_month", "full_previous_month"] = (
        "full_previous_month"
    )


class FixedCostSummaryResponse(BaseModel):
    expense_total: int
    fixed_total: int
    variable_total: int
    fixed_ratio: float | None
    essential_fixed_total: int
    discretionary_fixed_total: int
    essential_variable_total: int
    discretionary_variable_total: int
    required_spend_total: int
    discretionary_spend_total: int
    unclassified_total: int
    unclassified_count: int
    necessity_unclassified_total: int = 0
    necessity_unclassified_count: int = 0


class FixedCostTrendItem(BaseModel):
    period: str
    expense_total: int
    fixed_total: int
    variable_total: int
    essential_fixed_total: int
    discretionary_fixed_total: int
    essential_variable_total: int
    discretionary_variable_total: int
    required_spend_total: int
    discretionary_spend_total: int
    unclassified_total: int
    unclassified_count: int
    necessity_unclassified_total: int = 0
    necessity_unclassified_count: int = 0
    fixed_ratio: float | None


class FixedCostTrendResponse(BaseModel):
    items: list[FixedCostTrendItem]


class MerchantSpendItem(BaseModel):
    merchant: str
    amount: int
    count: int
    avg_amount: float
    last_seen_at: datetime


class MerchantSpendResponse(BaseModel):
    items: list[MerchantSpendItem]


# P1 ── payment-method-patterns
class PaymentMethodPatternItem(BaseModel):
    payment_method: str
    total_amount: int
    transaction_count: int
    avg_amount: int
    pct_of_total: float | None


class PaymentMethodPatternsResponse(BaseModel):
    items: list[PaymentMethodPatternItem]


# P1 ── income-stability
class IncomeMonthlyItem(BaseModel):
    period: str
    income: int


class IncomeStabilityResponse(BaseModel):
    items: list[IncomeMonthlyItem]
    avg: int
    stdev: float | None
    coefficient_of_variation: float | None
    comparison_mode: str
    reference_date: date
    is_partial_period: bool
    assumptions: str


# P1 ── recurring-payments
class RecurringPaymentItem(BaseModel):
    merchant: str
    category: str
    avg_amount: int
    interval_type: str
    avg_interval_days: float
    occurrences: int
    confidence: float
    last_date: date
    recurring_payment_kind: str | None
    installment_count: int
    monthly_recurring_count: int
    not_recurring_count: int
    unclassified_count: int
    transaction_ids: list[int]
    last_charge_date: date | None = None
    net_amount: int = 0
    activity_status: Literal[
        "active_candidate", "historical", "irregular", "non_positive", "not_recurring"
    ] = "historical"


class RecurringPaymentsResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[RecurringPaymentItem]
    reference_date: date | None = None
    activity: Literal["all", "active", "history"] = "all"
    recent_days: int = 90
    assumptions: str


# P1 ── spending-anomalies
class SpendingAnomalyItem(BaseModel):
    period: str
    category: str
    amount: int
    baseline_avg: int
    delta_pct: float | None
    delta_pct_raw: float | None = None
    delta_pct_display: float | None = None
    delta_display_capped: bool = False
    baseline_quality: str = "sufficient"
    anomaly_mode: str = "standard"
    direction: Literal["increase", "decrease"] = "increase"
    anomaly_score: float
    reason: str


class SpendingAnomaliesResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[SpendingAnomalyItem]
    comparison_mode: str
    reference_date: date
    is_partial_period: bool
    assumptions: str


class DiscretionaryVelocityResponse(BaseModel):
    period: str
    as_of_date: date
    month_progress_ratio: float
    discretionary_spend: int
    baseline_monthly_spend: int
    baseline_spend_at_same_progress: int
    velocity_ratio: float | None
    risk_level: str
    confidence: str
    classification_coverage_ratio: float | None
    unclassified_spend: int
    income_basis: str | None
    reasons: list[str]
    assumptions: list[str]


class PurchaseGateCandidateItem(BaseModel):
    candidate_type: str
    candidate_types: list[str]
    transaction_id: int
    candidate_key: str
    date: date
    merchant: str
    amount: int
    category: str
    signals: dict[str, int | float | str | bool]
    risk_level: str
    review_priority: str
    confidence: str
    suggested_review_window: str
    reasons: list[str]
    assumptions: list[str]
    review_status: str
    possible_cancellation: bool = False
    cancellation_evidence_transaction_ids: list[int] = Field(default_factory=list)
    review_memo: str | None = None
    reviewed_at: datetime | None = None
    cooldown_until: datetime | None = None
    review_timing: Literal["post_transaction"] = "post_transaction"
    candidate_purpose: Literal["future_friction_rule_candidate"] = (
        "future_friction_rule_candidate"
    )
    future_friction_suggestion: dict[str, object] | None = None


class PurchaseGateCandidatesResponse(BaseModel):
    total: int
    page: int
    per_page: int
    items: list[PurchaseGateCandidateItem]
    start_date: date | None = None
    end_date: date | None = None
    assumptions: list[str]


PurchaseGateReviewStatus = Literal[
    "pending",
    "reviewed",
    "ignored",
    "snoozed",
    "dismissed",
]


class PurchaseGateReviewPatchRequest(BaseModel):
    review_status: PurchaseGateReviewStatus
    memo: str | None = None
    cooldown_days: int | None = None


class PurchaseGateReviewResponse(BaseModel):
    candidate_key: str
    candidate_type: str
    transaction_id: int
    review_status: PurchaseGateReviewStatus
    memo: str | None = None
    reviewed_at: datetime | None = None
    cooldown_until: datetime | None = None

File: frontend/src/types/analytics.ts
export interface MonthlyCashflowItem {
  period: string
  income: number
  expense: number
  transfer: number
  net_cashflow: number
  savings_rate: number | null
}

export interface MonthlyCashflowResponse {
  items: MonthlyCashflowItem[]
}

export interface CategoryMoMItem {
  period: string
  previous_period: string
  category: string
  current_amount: number
  previous_amount: number
  delta_amount: number
  delta_pct: number | null
}

export interface CategoryMoMResponse {
  items: CategoryMoMItem[]
}

export interface CategoryMoMQuery {
  months?: number
  start_month?: string
  end_month?: string
  base_month?: string
}

export interface FixedCostSummaryResponse {
  expense_total: number
  fixed_total: number
  variable_total: number
  fixed_ratio: number | null
  essential_fixed_total: number
  discretionary_fixed_total: number
  essential_variable_total: number
  discretionary_variable_total: number
  required_spend_total: number
  discretionary_spend_total: number
  unclassified_total: number
  unclassified_count: number
}

export interface FixedCostTrendItem {
  period: string
  expense_total: number
  fixed_total: number
  variable_total: number
  essential_fixed_total: number
  discretionary_fixed_total: number
  essential_variable_total: number
  discretionary_variable_total: number
  required_spend_total: number
  discretionary_spend_total: number
  unclassified_total: number
  unclassified_count: number
  fixed_ratio: number | null
}

export interface FixedCostTrendResponse {
  items: FixedCostTrendItem[]
}

export interface MerchantSpendItem {
  merchant: string
  amount: number
  count: number
  avg_amount: number
  last_seen_at: string
}

export interface MerchantSpendResponse {
  items: MerchantSpendItem[]
}

export interface IncomeStabilityResponse {
  items: Array<{ period: string; income: number }>
  avg: number
  stdev: number | null
  coefficient_of_variation: number | null
  comparison_mode: 'closed' | 'partial'
  reference_date: string
  is_partial_period: boolean
  assumptions: string
}

export interface RecurringPaymentItem {
  merchant: string
  category: string
  avg_amount: number
  interval_type: string
  avg_interval_days: number
  occurrences: number
  confidence: number
  last_date: string
  recurring_payment_kind: 'installment' | 'monthly_recurring' | 'not_recurring' | null
  installment_count: number
  monthly_recurring_count: number
  not_recurring_count: number
  unclassified_count: number
  transaction_ids: number[]
}

export interface RecurringPaymentsResponse {
  total: number
  page: number
  per_page: number
  items: RecurringPaymentItem[]
  assumptions: string
}

export interface SpendingAnomalyItem {
  period: string
  category: string
  amount: number
  baseline_avg: number
  delta_pct: number | null
  delta_pct_raw?: number | null
  delta_pct_display?: number | null
  delta_display_capped?: boolean
  baseline_quality?: string
  anomaly_mode?: string
  anomaly_score: number
  reason: string
}

export interface SpendingAnomaliesResponse {
  total: number
  page: number
  per_page: number
  items: SpendingAnomalyItem[]
  comparison_mode: 'closed' | 'partial'
  reference_date: string
  is_partial_period: boolean
  assumptions: string
}

export interface IncomeStabilityQuery {
  end_date?: string
}

export interface SpendingAnomaliesQuery {
  page?: number
  per_page?: number
  end_date?: string
}

export type AnalyticsRiskLevel = 'low' | 'watch' | 'warning' | 'high' | 'critical'

export interface DiscretionaryVelocityResponse {
  period: string
  as_of_date: string
  month_progress_ratio: number
  discretionary_spend: number
  baseline_spend_at_same_progress: number
  velocity_ratio: number | null
  risk_level: AnalyticsRiskLevel
  review_priority?: number | null
  confidence: number
  reasons: string[]
  assumptions: string[]
  unclassified_spend: number
  classification_coverage_ratio: number
  income_basis?: 'observed' | 'estimated' | string | null
}

export interface DiscretionaryVelocityQuery {
  as_of_date?: string
}

export type PurchaseGateCandidateType =
  | 'large_oneoff'
  | 'new_merchant'
  | 'merchant_spike'
  | 'discretionary_spike'

export type PurchaseGateCandidateStatus =
  | 'pending'
  | 'reviewed'
  | 'ignored'
  | 'snoozed'
  | 'dismissed'

export interface PurchaseGateCandidateItem {
  candidate_type: PurchaseGateCandidateType
  candidate_types: string[]
  transaction_id: number
  candidate_key: string
  date: string
  merchant: string
  amount: number
  category: string
  signals: Record<string, number | string | boolean>
  risk_level: AnalyticsRiskLevel
  review_priority: string
  confidence: string
  suggested_review_window: string
  review_status: PurchaseGateCandidateStatus
  review_memo?: string | null
  reviewed_at?: string | null
  cooldown_until?: string | null
  review_timing?: 'post_transaction'
  candidate_purpose?: 'future_friction_rule_candidate'
  future_friction_suggestion?: {
    condition?: Record<string, unknown>
    action?: string
  } | null
  reasons: string[]
  assumptions: string[]
}

export interface PurchaseGateCandidatesResponse {
  total: number
  page: number
  per_page: number
  items: PurchaseGateCandidateItem[]
  assumptions: string[]
}

export interface PurchaseGateCandidatesQuery {
  status?: PurchaseGateCandidateStatus
  limit?: number
}

export interface PurchaseGateReviewPatchRequest {
  review_status: PurchaseGateCandidateStatus
  memo?: string | null
  cooldown_days?: number | null
}
